### ISSUE TITLE
no-unused-vars should report all unused function parameters that match the configured setting

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#9909]
I'm creating this as an attempt to refocus the discussion in #9750 into a specific proposal I have (although I think @g-plane would agree with me as well).

**What rule do you want to change?**

no-unused-vars

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

New default behavior for `args: "after-used"` option: Report all unused parameters which occur after the last used parameter. This would be done as 1 lint message per unused parameter.

Note that this is not really a breaking change:
* When the user has no unused parameters after the last used parameter, the rule will continue not to report any violations.
* When the user has at least one unused parameter after this last used parameter, we currently report 1 parameter at a time.
    * When the user removes the unused parameter, we report the next unused parameter. Net effect for the user is to (eventually) see lint messages for all unused parameters after the last used parameter.
    * This proposal is just to report them all at once.

**Please provide some example code that this change will affect:**

```js
/* eslint no-unused-vars: ["error", { "args": "after-used" }] */
function (a, b, c, d, e) {
    c();
}
```

For more examples, see [demo](https://eslint.org/demo/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC12YXJzOltlcnJvciwgeyBhcmdzOiBhZnRlci11c2VkIH1dICovXHJcblxyXG5mdW5jdGlvbiBvbmUoYSwgYiwgYykge1xyXG4gICAgLy8gbm9uZSB1c2VkOiBTaG91bGQgcmVwb3J0IGEsIGIsIGFuZCBjIChldmVuIHdpdGggdGhlIFwiYWZ0ZXItdXNlZFwiIHNldHRpbmcpXHJcbn1cclxuXHJcbmZ1bmN0aW9uIHR3byhhLCBiLCBjLCBkLCBlKSB7XHJcbiAgICBkb1NvbWV0aGluZyhjKTtcclxuICAgIC8vIGEvYi9kL2UgdW51c2VkOiBTaG91bGQgcmVwb3J0IGEvYi9kL2UgaW4gXCJhbGxcIiBzZXR0aW5nLCBhbmQgZC9lIGluIFwiYWZ0ZXItdXNlZFwiIHNldHRpbmcpXHJcbn1cclxuXHJcbmZ1bmN0aW9uIHRocmVlKGEsIGIsIGMpIHtcclxuICAgIGRvU29tZXRoaW5nKGMpO1xyXG4gICAgLy8gYS9iIHVudXNlZDogU2hvdWxkIHJlcG9ydCBhL2IgaW4gXCJhbGxcIiBzZXR0aW5nLCBhbmQgc2hvdWxkIG5vdCByZXBvcnQgaW4gXCJhZnRlci11c2VkXCIgc2V0dGluZylcclxufVxyXG5cclxub25lKCk7XHJcbnR3bygpO1xyXG50aHJlZSgpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6Niwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**What does the rule currently do for this code?**

For code block above: Reports that `e` is unused.

For demo: See lint results.

**What will the rule do after it's changed?**

For code block above: Reports that `d`, and `e` are all unused.

* `a`, `b`, `d`, and `e` are unused.
* But `d` and `e` are after the last used parameter (`c`).
* So report both `d` and `e`.

For demo: See comments in code.

Note: `args: "all"` works correctly and needs no changes.

--------

I'll champion.

If there's a strong reason for reporting only one of the unused arguments (and no, "the docs say so" doesn't count, that's circular reasoning), that can be expressed here as part of a :-1: to this proposal.

**Comment by g-plane** [comment]
LGTM

**Comment by platinumazure** [comment]
@eslint/eslint-team Friendly ping: Could we evaluate this proposal to change no-unused-vars and make the UX a little better? Or, for those who have expressed opposition, could we hash that out here? Thanks!

**Comment by not-an-aardvark** [comment]
@platinumazure The issue has now been accepted.

**Comment by platinumazure** [comment]
Thanks everyone!

@g-plane Now that this is accepted and we have a roadmap, is it okay if I close your original issue? Thanks!

**Comment by platinumazure** [comment]
I've added "breaking" label for now, but I disagree that this is a breaking change. I've explained why in the initial post, but here it is again for convenience:

> Note that this is not really a breaking change:
> * When the user has no unused parameters after the last used parameter, the rule will continue not to report any violations.
> * When the user has at least one unused parameter after this last used parameter, we currently report 1 parameter at a time.
>     * When the user removes the unused parameter, we report the next unused parameter. Net effect for the user is to (eventually) see lint messages for all unused parameters after the last used parameter.
>     * This proposal is just to report them all at once.

The overall number of errors is actually unchanged; it's just reported over one linting runs. And-- this is the important part-- there is no situation where 0 issues are currently reported, and >0 issues are reported under this proposal.

Using this example as a reference:

```js
/* eslint no-unused-vars: ["error", { "args": "after-used" }] */
function (a, b, c, d, e) {
    c();
}
```

In my example, we instead go from 1 issue reported in 2 lint runs (`e` reported, then `d` when `e` is removed), to 2 issues reported (`d` and `e` in the same run).

In this example:

```js
/* eslint no-unused-vars: ["error", { "args": "after-used" }] */
function (a, b, c) {
    c();
}
```

No report is generated right now, and no report will be generated with this proposal.

So since there are no new linting errors introduced where previously there were no linting errors, I don't think this is a breaking change.

**Comment by not-an-aardvark** [comment]
> I know this is already accepted, so I won't make a big fuss, however, has anyone researched why "after-used" currently behaves like this? I'm guessing something like this issue has been reported before and there's a reason why the current behavior exists. It would be good to get that background before moving forward.

Based on the git history, it seems like this behavior was introduced in https://github.com/eslint/eslint/pull/1219. There wasn't any discussion about the behavior change, so my best guess is that it was unintentional.

**Comment by nzakas** [comment]
Okay, I think I'm confused about some things, so please bear with me.

@platinumazure Will there be an individual lint message for each argument that is not used after one that has been used? Or will there be one lint message that says something like "d and e are unused"? (I think it should be the former, but the wording in your proposal isn't clear to me.)

@not-an-aardvark Are you saying that previously, the "after-used" option worked the way that @platinumazure is now proposing?

In any case, the proposal says this change introduces more warnings and is an enhancement, so it must be a breaking change by our definition. If, however, we're saying that this is a bug as the behavior was inadvertently changed at some point in the past and this proposal restores the original behavior, then I think there's a good argument that it's not a breaking change. 

There's still an open question of whether or not people are relying on the current behavior. We can attempt to logically reason that they are not, but attempts at such things in the past have not turned out accurate. That's why I think adding a "last-only" option that covers the current behavior might be worthwhile.

**Comment by platinumazure** [comment]
> Will there be an individual lint message for each argument that is not used after one that has been used? Or will there be one lint message that says something like "d and e are unused"? (I think it should be the former, but the wording in your proposal isn't clear to me.)

One lint message for each unused argument. I'll update the proposal to make that more clear.

> In any case, the proposal says this change introduces more warnings and is an enhancement, so it must be a breaking change by our definition.

That's the thing, though, this doesn't really introduce more warnings. It just reports all the warnings in one lint run, instead of multiple; and more importantly, it does not introduce warnings in cases that reported no warnings. So no users will be broken by this change. I think this is true whether we consider this a bugfix or an enhancement.

> There's still an open question of whether or not people are relying on the current behavior. We can attempt to logically reason that they are not, but attempts at such things in the past have not turned out accurate. That's why I think adding a "last-only" option that covers the current behavior might be worthwhile.

That's fair, and I'm not opposed to adding a "last-only" option. But does that need to be part of this proposal? Or could we tackle that separately?

In any case, due to the above (i.e., my suggestion that the change is not really breaking), it seems to me that if we get this wrong, the impact is fairly minimal and we could add a "last-only" option if users request it.

**Comment by not-an-aardvark** [comment]
> @not-an-aardvark Are you saying that previously, the "after-used" option worked the way that @platinumazure is now proposing?

Yes, I think that's correct based on looking at the old versions of the code.

> In any case, the proposal says this change introduces more warnings and is an enhancement, so it must be a breaking change by our definition.

I think this might be coming from an ambiguity in our definition. To be clear, this increases the *number* of warnings on some lint runs, but it only does this in cases where at least one warning would have been reported anyway. In other words, there are no cases in which a warning would start to be reported where no warnings were reported before.

There was a similar case in https://github.com/eslint/eslint/pull/7095#issuecomment-246195723 when we updated `quote-props` to report individual errors for each property, rather than reporting the object as a whole. We ended up calling that a semver-minor change.

**Comment by g-plane** [comment]
I agree with adding `last-only` option. However I suggest that keep `after-used` as default(it won't bring breaking changes). If using `after-used` (this proposal) as default, ESLint can report all unused arguments that _after last used_.

Initially, the reason why I opened #9750 is that I didn't know there was a option `args: all` at that time and ESLint just reported the last unused argument. Though **NOW** I know it is technically correct, that behavior confuses me and I thought it may be a bug **BEFORE**. If applying this proposal, those users like me won't feel confused.

So I agree with this proposal and keep `after-used` as default. Any discussions are welcome.

**Comment by not-an-aardvark** [comment]
**TSC Summary**: This issue proposes changing the behavior of `after-used` in `no-unused-vars` to report all parameters after the final used parameter, rather than only reporting the last parameter, if there are multiple unused parameters at the end of a parameter list.

**TSC Question**: Should this be considered a breaking change? Should we add a `last-only` option for users who want to retain the current behavior?

**Comment by kaicataldo** [comment]
The TSC discussed this at today's meeting. The resolution is to accept this as a semver-minor bugfix. We will not add a `last-only` option right now.

**Comment by platinumazure** [comment]
Switching "enhacement" label to "bug" based on the TSC decision.

**Comment by nzakas** [comment]
Sorry, couldn't get out of bed yesterday so wasn't able to respond. I see the TSC already voted on this, so I'd just like to include the answers to the points brought up here for consideration in other issues going forward.

1. Changing the number of warnings a rule outputs on a given piece of code is a breaking change because we don't know that people aren't relying on the number of warnings to trigger something. In this context, I mean "breaking change" as in it will change how people are using ESLint in an incompatible way. I saw in the TSC notes that `-max-warnings` was mentioned, and that's part of it. We also don't know if people are just outputting the results raw and then doing something with the count. This is explicitly why the semantic versioning policy mentions *number* of warnings, not when the warnings occur. Now, if we are accidentally missing that should have been warned or used to be warned, then it's  a bug fix and it's semver-minor because we made a mistake that needs to be fixed and so it's worth breaking people's experiences because they might be missing something important. 

2. If there's a regression, it's important to note *how long ago* the behavior regressed. A bug fix that introduces more warnings should, in my opinion, be a breaking change if the behavior has been around long enough that people might reasonably be relying on the broken behavior. In this case, it looks like @not-an-aardvark tracked the change to #1219, which was merged over three and a half years ago. This was my point in suggesting adding "last-only". If we've had this behavior for the past three and a half years, users are going to be disrupted with a change and we need to give them a way to get back to what they had been using.

I'm sorry I didn't explain this well enough earlier. I'm not sure it would have made a difference, but I really think that with a user base as large as ESLint's that we need to be more conservative with how we introduce changes that are on the edge of being a breaking change.

**Comment by platinumazure** [comment]
Looks like nobody has had a chance to really work on this since the discussion, so I think we could reasonably just call it a breaking change and get it out with 5.0.0.

**Comment by platinumazure** [comment]
I've added the "breaking" label to this issue, and I've created PR #10119 for implementing this change (also labeled as "breaking").
